File: csharp/src/Training.Workshop/M5/S12BridgeMethods/Step2/StudentRule.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M5.S12BridgeMethods.Step2;

/// <summary>Krok 2: reguła implementuje tylko metodę dla swojego biletu - typ i "most" daje kontrakt.</summary>
public sealed class StudentRule : IPriceRule<StudentTicket>
{
    public Money Apply(StudentTicket ticket)
    {
        return ticket.BasePrice.Minus(ticket.BasePrice.Percent(25));
    }
}

File: typescript/src/workshop/m5/s12_bridgemethods/start/StudentRule.ts
import type { Money } from '../../../shared/Money.js';
import type { StudentTicket } from '../StudentTicket.js';

/** Start: reguła studencka (-25%) - też bez wspólnego typu. */
export class StudentRule {
  apply(ticket: StudentTicket): Money {
    return ticket.basePrice().minus(ticket.basePrice().percent(25));
  }
}
